languages/python/sqlalchemy-oso/tests/test_sqlalchemy.py

```diff
@@ -0,0 +1,331 @@
+"""Test hooks & SQLAlchemy API integrations."""
+import pytest
+
+from sqlalchemy.orm import aliased, sessionmaker
+
+from sqlalchemy_oso.hooks import (
+    authorize_query,
+    enable_hooks,
+    make_authorized_query_cls,
+    authorized_sessionmaker,
+)
+
+from .models import *
+from .conftest import print_query
+
+
+def log_queries():
+    import logging
+
+    logging.basicConfig()
+    logging.getLogger("sqlalchemy.engine").setLevel(logging.INFO)
+
+
+def test_authorize_query_no_access(session, oso, fixture_data):
+    query = session.query(Post)
+
+    authorized = authorize_query(query, lambda: oso, lambda: "user", lambda: "action")
+    assert authorized.count() == 0
+
+
+def test_authorize_query_basic(session, oso, fixture_data):
+    # TODO: copied from test_authorize_model_basic
+    oso.load_str('allow("user", "read", post: Post) if post.access_level = "public";')
+    oso.load_str('allow("user", "write", post: Post) if post.access_level = "private";')
+    oso.load_str('allow("admin", "read", post: Post);')
+    oso.load_str(
+        'allow("moderator", "read", post: Post) if '
+        '(post.access_level = "private" or post.access_level = "public") and '
+        "post.needs_moderation = true;"
+    )
+
+    query = session.query(Post)
+    authorized = authorize_query(query, lambda: oso, lambda: "user", lambda: "read")
+
+    print_query(authorized)
+    assert authorized.count() == 5
+    assert authorized.all()[0].contents == "foo public post"
+    assert authorized.all()[0].id == 0
+
+    posts = authorize_query(query, lambda: oso, lambda: "user", lambda: "write")
+
+    assert posts.count() == 4
+    assert posts.all()[0].contents == "foo private post"
+    assert posts.all()[1].contents == "foo private post 2"
+
+    posts = authorize_query(query, lambda: oso, lambda: "admin", lambda: "read")
+    assert posts.count() == 9
+
+    posts = authorize_query(query, lambda: oso, lambda: "moderator", lambda: "read")
+    print_query(posts)
+    assert posts.all()[0].contents == "private for moderation"
+    assert posts.all()[1].contents == "public for moderation"
+
+    posts = authorize_query(query, lambda: oso, lambda: "guest", lambda: "read")
+    assert posts.count() == 0
+
+
+def test_authorize_query_multiple_entities(session, oso, fixture_data):
+    # TODO: copied from test_authorize_model_basic
+    oso.load_str('allow("user", "read", post: Post) if post.access_level = "public";')
+    oso.load_str('allow("user", "write", post: Post) if post.access_level = "private";')
+    oso.load_str('allow("admin", "read", post: Post);')
+    oso.load_str(
+        'allow("moderator", "read", post: Post) if '
+        '(post.access_level = "private" or post.access_level = "public") and '
+        "post.needs_moderation = true;"
+    )
+
+    query = session.query(Post.contents, Post.id)
```

Comment:
Could this test + the one above be extracted into a shared setup where we pass in the query? Since that seems to be the only difference between the two tests

Reply:
Fixed up 510ff36dc3b5eac8c530aa9cf55643321ece987b

Reply:
nicely done